File: rb/test/episode_entity_test.rb
# Episode entity test

require "minitest/autorun"
require "json"
require_relative "../Tvmaze_sdk"
require_relative "runner"

class EpisodeEntityTest < Minitest::Test
  def test_create_instance
    testsdk = TvmazeSDK.test(nil, nil)
    ent = testsdk.Episode(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "episode" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = TvmazeSDK.test(seed, nil)
    seen = base.Episode(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = TvmazeConfig.make_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = TvmazeSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.Episode(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = episode_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "episode." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set TVMAZE_TEST_EPISODE_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    episode_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.episode")))
    episode_ref01_data = nil
    if episode_ref01_data_raw.length > 0
      episode_ref01_data = Helpers.to_map(episode_ref01_data_raw[0][1])
    end

    # LIST
    episode_ref01_ent = client.Episode(nil)
    episode_ref01_match = {
      "show_id" => setup[:idmap]["show01"],
    }

    episode_ref01_list_result = episode_ref01_ent.list(episode_ref01_match, nil)
    assert episode_ref01_list_result.is_a?(Array)

    # LOAD
    episode_ref01_match_dt0 = {
      "id" => episode_ref01_data["id"],
    }
    episode_ref01_data_dt0_loaded = episode_ref01_ent.load(episode_ref01_match_dt0, nil)
    episode_ref01_data_dt0_load_result = Helpers.to_map(episode_ref01_data_dt0_loaded)
    assert !episode_ref01_data_dt0_load_result.nil?
    assert_equal episode_ref01_data_dt0_load_result["id"], episode_ref01_data["id"]

  end
end

def episode_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "episode", "EpisodeTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = TvmazeSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["episode01", "episode02", "episode03", "season01", "season02", "season03", "show01", "show02", "show03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["TVMAZE_TEST_EPISODE_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "TVMAZE_TEST_EPISODE_ENTID" => idmap,
    "TVMAZE_TEST_LIVE" => "FALSE",
    "TVMAZE_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["TVMAZE_TEST_EPISODE_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["TVMAZE_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = TvmazeSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["TVMAZE_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["TVMAZE_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: php/test/EpisodeEntityTest.php
<?php
declare(strict_types=1);

// Episode entity test

require_once __DIR__ . '/../tvmaze_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class EpisodeEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = TvmazeSDK::test(null, null);
        $ent = $testsdk->Episode(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = episode_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "episode." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set TVMAZE_TEST_EPISODE_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $episode_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.episode")));
        $episode_ref01_data = null;
        if (count($episode_ref01_data_raw) > 0) {
            $episode_ref01_data = Helpers::to_map($episode_ref01_data_raw[0][1]);
        }

        // LIST
        $episode_ref01_ent = $client->Episode(null);
        $episode_ref01_match = [
            "show_id" => $setup["idmap"]["show01"],
        ];

        $episode_ref01_list_result = $episode_ref01_ent->list($episode_ref01_match, null);
        $this->assertIsArray($episode_ref01_list_result);

        // LOAD
        $episode_ref01_match_dt0 = [
            "id" => $episode_ref01_data["id"],
        ];
        $episode_ref01_data_dt0_loaded = $episode_ref01_ent->load($episode_ref01_match_dt0, null);
        $episode_ref01_data_dt0_load_result = Helpers::to_map($episode_ref01_data_dt0_loaded);
        $this->assertNotNull($episode_ref01_data_dt0_load_result);
        $this->assertEquals($episode_ref01_data_dt0_load_result["id"], $episode_ref01_data["id"]);

    }
}

function episode_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/episode/EpisodeTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = TvmazeSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["episode01", "episode02", "episode03", "season01", "season02", "season03", "show01", "show02", "show03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("TVMAZE_TEST_EPISODE_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "TVMAZE_TEST_EPISODE_ENTID" => $idmap,
        "TVMAZE_TEST_LIVE" => "FALSE",
        "TVMAZE_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["TVMAZE_TEST_EPISODE_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["TVMAZE_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new TvmazeSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["TVMAZE_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["TVMAZE_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
